nengo/spa/module.py

```diff
@@ -25,3 +51,159 @@ def on_add(self, spa):
+        if ex_type is not None:
+            # re-raise the exception that triggered this __exit__
+            return False
+
+        module_list = frozenset(self._modules.values())
+        for net in self.networks:
+            # Since there are no attributes to distinguish what's been added
+            # and what hasn't, we have to ask the network
+            if isinstance(net, Module) and (net not in module_list):
+                raise SpaModuleError("%s must be set as an attribute of "
+                                     "a SPA network" % (net))
+
+    def get_module(self, name, strip_output=False):
+        """Return the module for the given name."""
+        try:
+            components = name.split('.', 1)
+            if len(components) > 1:
+                head, tail = components
+                return self._modules[head].get_module(
+                    tail, strip_output=strip_output)
+            else:
+                if name in self._modules:
+                    return self._modules[name]
+                elif strip_output and (
+                        name in self.inputs or name in self.outputs):
+                    return self
+                else:
+                    raise KeyError
+        except KeyError:
+            raise SpaModuleError("Could not find module %r" % name)
+
+    def get_module_input(self, name):
+        """Return the object to connect into for the given name.
+
+        The name will be either the same as a module, or of the form
+        <module_name>.<input_name>.
+        """
+        try:
+            components = name.split('.', 1)
+            if len(components) > 1:
+                head, tail = components
+                return self._modules[head].get_module_input(tail)
+            else:
+                if name in self.inputs:
+                    return self.inputs[name]
+                elif name in self._modules:
+                    return self._modules[name].get_module_input('default')
+                else:
+                    components = name.rsplit('_', 1)
+                    if len(components) > 1:
+                        head, tail = components
+                        inp = self._modules[head].get_module_input(tail)
+                        warnings.warn(DeprecationWarning(
+                            "Underscore notation for inputs and outputs is "
+                            "deprecated. Use dot notation <module>.<name> "
+                            "instead."))
+                        return inp
+                    else:
+                        raise KeyError
+        except KeyError:
+            raise SpaModuleError("Could not find module input %r" % name)
+
+    def get_module_inputs(self):
+        for name, module in iteritems(self._modules):
+            for inp in module.inputs:
+                if inp == 'default':
+                    yield name
+                else:
+                    yield '%s_%s' % (name, inp)
```

Comment:
Tricky....  I'm very tempted to just declare that these shouldn't have been part of the public interface in the first place, and so I'm not all that worried about removing them.  I think it's less of a change than the change that's requiring replacing "_" with "." in Spaun.  That actually breaks existing code (but in an easy-to-fix way that results in better code), whereas removing this is only potentially breaking things.  So I don't think what we decide here would be the cause of pushing off this PR, and so let's just get rid of the functions.  You're quite right that they can be inspected in other ways, and the approach used in this function doesn't make sense in hierarchical SPA.